sdk/core/azure_core_amqp/src/connection.rs

```diff
@@ -0,0 +1,315 @@
+// Copyright (c) Microsoft Corp. All Rights Reserved.
+// cspell: words amqp sasl
+
+use super::value::{AmqpOrderedMap, AmqpSymbol, AmqpValue};
+use azure_core::error::Result;
+use std::fmt::Debug;
+use time::Duration;
+use url::Url;
+
+pub struct AmqpConnectionOptions {
+    pub(crate) max_frame_size: Option<u32>,
+    pub(crate) channel_max: Option<u16>,
+    pub(crate) idle_timeout: Option<Duration>,
+    pub(crate) outgoing_locales: Option<Vec<String>>,
+    pub(crate) incoming_locales: Option<Vec<String>>,
+    pub(crate) offered_capabilities: Option<Vec<AmqpSymbol>>,
+    pub(crate) desired_capabilities: Option<Vec<AmqpSymbol>>,
+    pub(crate) properties: Option<AmqpOrderedMap<AmqpSymbol, AmqpValue>>,
+    pub(crate) buffer_size: Option<usize>,
+}
+
+impl AmqpConnectionOptions {
+    pub fn builder() -> builders::AmqpConnectionOptionsBuilder {
+        builders::AmqpConnectionOptionsBuilder::new()
+    }
+}
+
+pub trait AmqpConnectionTrait {
+    #[allow(unused_variables)]
+    fn open(
+        &self,
+        name: impl Into<String>,
+        url: Url,
+        options: Option<AmqpConnectionOptions>,
+    ) -> impl std::future::Future<Output = Result<()>> {
+        async {
+            unimplemented!();
+        }
+    }
+    fn close(&self) -> impl std::future::Future<Output = Result<()>> {
+        async {
+            unimplemented!();
+        }
+    }
+}
+
+#[cfg(all(feature = "enable-fe2o3-amqp", not(target_arch = "wasm32")))]
+type ConnectionImplementation = super::fe2o3::connection::Fe2o3AmqpConnection;
+
+#[cfg(any(not(feature = "enable-fe2o3-amqp"), target_arch = "wasm32"))]
+type ConnectionImplementation = super::noop::NoopAmqpConnection;
+
+#[derive(Debug, Default)]
+pub(crate) struct AmqpConnectionImpl<T>(pub(crate) T);
+
+impl<T> AmqpConnectionImpl<T>
+where
+    T: AmqpConnectionTrait,
+{
+    pub(crate) fn new(connection: T) -> Self {
+        Self(connection)
+    }
+}
+
+#[derive(Debug, Default)]
+pub struct AmqpConnection(pub(crate) AmqpConnectionImpl<ConnectionImplementation>);
+
+impl AmqpConnectionTrait for AmqpConnection {
+    fn open(
+        &self,
+        name: impl Into<String>,
+        url: Url,
+        options: Option<AmqpConnectionOptions>,
+    ) -> impl std::future::Future<Output = Result<()>> {
+        self.0 .0.open(name, url, options)
+    }
+    fn close(&self) -> impl std::future::Future<Output = Result<()>> {
+        self.0 .0.close()
+    }
+}
+
+impl AmqpConnection {
+    pub fn new() -> Self {
+        let connection = ConnectionImplementation::new();
+
+        Self(AmqpConnectionImpl::new(connection))
+    }
+}
+
+pub mod builders {
+    use super::*;
+    pub struct AmqpConnectionOptionsBuilder {
+        options: AmqpConnectionOptions,
+    }
+
+    impl AmqpConnectionOptionsBuilder {
+        pub(super) fn new() -> Self {
+            Self {
+                options: AmqpConnectionOptions {
```

Comment:
Tip: derive `Default` on `AmqpConnectionOptions` - since everything is an `Option<T>` it'll work - and just do: `options: Default::default(),` here.

Reply:
It'll be more generally useful anyway if it impls `Default`.

Reply:
Oooh, nice. Love that one.